File: android/src/main/java/com/enso/idnowautoident/IdNowAutoIdentModule.kt
package com.enso.idnowautoident

import de.idnow.core.IDnowConfig
import de.idnow.core.IDnowResult
import de.idnow.core.IDnowSDK
import de.idnow.core.IDnowSDK.IDnowResultListener
import expo.modules.kotlin.Promise
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import org.json.JSONObject


data class IdentResult(val status: String, val desResType: IDnowResult.ResultType) {
  private val description: String = when(desResType) {
    IDnowResult.ResultType.CANCELLED -> "CANCELLED"
    IDnowResult.ResultType.FINISHED -> "FINISHED"
    IDnowResult.ResultType.ERROR -> "ERROR"
    else -> "UNKNOWN"
  }

  fun getJSONResult(): JSONObject {
    val jsonObj = JSONObject()
    jsonObj.put("status", status)
    jsonObj.put("description", description)
    return jsonObj
  }
}

class IdNowAutoIdentModule : Module() {

  private lateinit var idnowSdk: IDnowSDK
  override fun definition() = ModuleDefinition {
    Name("IdNowAutoIdent")

    AsyncFunction("init") { language: String, promise: Promise ->
      val activity = appContext.activityProvider?.currentActivity
      val idnowConfig = IDnowConfig.Builder
              .getInstance()
              .withLanguage(language)
              .build()

      idnowSdk = IDnowSDK.getInstance()
      if (activity != null) {
        idnowSdk.initialize(activity, idnowConfig)
        promise.resolve("initialized sdk with language " + language)
      } else {
        promise.resolve("Failed to resolve idnow init activity is empty")
      }
    }

    AsyncFunction("start") { token: String, language: String, promise: Promise ->
      val listener = IDnowResultListener {it: IDnowResult ->
        val idkRes = IdentResult(it.statusCode, it.resultType)
        promise.resolve(idkRes.getJSONResult().toString())
      }
      idnowSdk.startIdent(token, listener)
    }
  }
}

File: ios/IdNowAutoIdentModule.swift
import ExpoModulesCore
import IDNowSDKCore


func convertToJsonString<T: Encodable>(data: T) -> String? {
    do {
        let jsonData = try JSONEncoder().encode(data)
        let jsonString = String(data: jsonData, encoding: .utf8)
        return jsonString
    } catch {
        print("Error converting to JSON: \(error)")
        return nil
    }
}

struct IdentResult : Codable {
    var status: String
    var description: String
}


func generateJSON(result: IdentResult) ->  String? {
    if let jsonString = convertToJsonString(data: result) {
        return jsonString
    } else {
        return "failed to pars object"
    }
}

func turnResultToString(result: IDNowSDK.IdentResult.type) -> String {
    if result == .CANCELLED {
        return "CANCELED"
    }
    else if result == .FINISHED {
        return "FINISHED"
    }
    return "ERROR"
    
}

public class IdNowAutoIdentModule: Module {

  public func definition() -> ModuleDefinition {
    Name("IdNowAutoIdent")
    AsyncFunction("start") { (token: String, preferredLanguage: String, promise: Promise) in
      if let rootViewController = UIApplication.shared.delegate?.window??.rootViewController {
        IDNowSDK.shared.start(token: token, preferredLanguage:preferredLanguage, fromViewController: rootViewController, listener:{(result: IDNowSDK.IdentResult.type, statusCode: IDNowSDK.IdentResult.statusCode, message: String) in
            let idkRes = IdentResult(status: turnResultToString(result: result), description: statusCode.description)
            promise.resolve(generateJSON(result: idkRes))
                
        })
      } else {
          promise.resolve("failed to identify rootViewControler")
      }
    }
  }
}
